### PR TITLE
fix: mse_vae training broken due to incorrect averaging

### PR DESCRIPTION
**Pull request by paulperet**
When training, the model generates blurry images and is unable to improve. Setting MSELoss size_average to False fixes the issue.

### CODE REVIEW

**Comment by orybkin** [comment]
Ah that's definitely wrong! size_average is incorrect because bigger images have more information so they should have a larger MSE loss.

It is possible that in some cases coincidentally size_average works better, which is related to balancing the two losses. However, it does not magically solve the issue - you will only get best performance if you are careful about balancing the losses. I suggest using the method from my paper to do that :)

**Comment by paulperet** [comment]
Yes I agree that some balancing needs to be done to achieve best performance, 

1. but I can't find any theoretical reason to average over the dimension of the input:

https://stats.stackexchange.com/questions/540092/how-do-we-get-to-the-mse-in-the-loss-function-for-a-variational-autoencoder

From what I understand the squared errors should be summed. This is similar to the "3.1. Gaussian Decoders"  of the sigma-vae paper, where the MSE is multiplied by D (the dimension of x).

2. Then why the kl term would be the sum over batch when the reconstruction loss would be the average over the batch? The most naive approach would be to sum the squared errors as I suggested instead of averaging (at least over the batch dimension)

I may be wrong but I noticed a lot of inconsistencies in how people implement the VAE in practice and I am trying my best to understand the field. I appreciated the quick follow up!

**Comment by orybkin** [comment]
Yes. Check the original 2014 VAE paper. The KL term is summed over latent dimensions and the MSE term is summed over pixels. This is because additional latent dimensions contain additional information and additional pixels also contain additional information. If you have more pixels, you want the MSE loss to be larger, and if you have have more latent dimensions that are not unit gaussian you want a larger KL divergence loss. 

Re batch - you can either sum or average over batch if your batches are all the same size. This is just a constant factor that will be removed by Adam. 

In practice, people often average over pixels and latent dimensions. This is wrong for the above reason and leads to bad balance. Sometimes it leads to a good balance, which is purely an accident.

**Comment by paulperet** [comment]
Yes so we agree on everything. I guess this is supposed to be an example of wrong practice inspired by the PyTorch example. Maybe this sentence from the README leads to confusion: "First, a VAE from the original PyTorch example repo that uses MSE loss.". It implies that the first example is directly copied from the PyTorch example. But their example uses reduction='sum' on the BCELoss, which is correct.

Maybe you could add a line in the code or README that specifically says that it is a common error found on internet (in practice)?

**Comment by orybkin** [comment]
Oh interesting. It's possible the example has sum now. At the time I wrote this I checked a large number of repos and most had average :)
